org.eclipse.dawnsci.analysis.dataset/src/org/eclipse/dawnsci/analysis/dataset/FunctionalUtils.java

```diff
@@ -0,0 +1,352 @@
+/*-
+ * Copyright 2017 Diamond Light Source Ltd.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.eclipse.dawnsci.analysis.dataset;
+
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.function.DoubleConsumer;
+import java.util.function.IntConsumer;
+import java.util.function.IntUnaryOperator;
+import java.util.function.LongConsumer;
+import java.util.stream.DoubleStream;
+import java.util.stream.IntStream;
+import java.util.stream.LongStream;
+import java.util.stream.StreamSupport;
+
+import org.eclipse.january.dataset.Dataset;
+import org.eclipse.january.dataset.DoubleDataset;
+import org.eclipse.january.dataset.IndexIterator;
+import org.eclipse.january.dataset.IntegerDataset;
+import org.eclipse.january.dataset.LongDataset;
+
+/**
+ * Provide static methods to use Datasets in functional ways 
+ */
+public class FunctionalUtils {
+
+	private FunctionalUtils() {
+	}
+
+	private static class DatasetIndexOperator implements IntUnaryOperator {
+
+		private int[] shape;
+		private int[] strides;
+		private int offset;
+		private int rank;
+
+		public DatasetIndexOperator(Dataset a) {
+			shape = a.getShape();
+			rank = shape.length;
+			strides = a.getStrides().clone();
+			offset = a.getOffset();
+		}
+
+		@Override
+		public int applyAsInt(int value) {
+			int index = offset;
+			for (int i = rank-1; value > 0 && i >= 0; i--) {
+				int s = shape[i];
+				int p = value % s;
+				if (p != 0) {
+					index += p * strides[i];
+				}
+				value /= s;
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Introduce a new variable instead of reusing the parameter "value". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS1226)